### PR TITLE
Problem: implement topology services from fty-rest topology applets

### PR DESCRIPTION
**Pull request by jeffdegott**
Solution: move codes, implement mailbox

### CODE REVIEW

File: src/topology/db/topology2.cc [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t5name ON (t5.id = t5name.id_asset_element AND t5name.keytag=\"name\")        "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t6name ON (t6.id = t6name.id_asset_element AND t6name.keytag=\"name\")        "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t7name ON (t7.id = t7name.id_asset_element AND t7name.keytag=\"name\")        "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t8name ON (t8.id = t8name.id_asset_element AND t8name.keytag=\"name\")        "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t9name ON (t9.id = t9name.id_asset_element AND t9name.keytag=\"name\")        "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t10name ON (t10.id = t10name.id_asset_element AND t10name.keytag=\"name\")    "
+        "    LEFT JOIN t_bios_asset_ext_attributes AS t11name ON (t11.id = t11name.id_asset_element AND t11name.keytag=\"name\")    "
+        "    INNER JOIN t_bios_asset_device_type AS v1 ON (v1.id_asset_device_type = t1.id_subtype) "
+        "  WHERE t1.name=:from ";
```
Comment (boricj) [inline src/topology/db/topology2.cc]: I weep for my profession...

File: src/topology/shared/location_helpers.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    json += "\"name\" : \"" + UTF8::escape (elem_names.second) + "\", ";
+    json += "\"id\" : \"" + UTF8::escape (name) + "\",";
+    json += "\"type\" : \"" + persist::typeid_to_type(type_id) + "\",";
+    if ( (type_id == persist::asset_type::DEVICE ) ||
+         (type_id == persist::asset_type::GROUP) ) {
+        json += "\"sub_type\" : \"" + utils::strip (type_name) + "\"";
+    }
+    else {
+        json += "\"sub_type\" : \"N_A\"";
+    }
```
Comment (boricj) [inline src/topology/shared/location_helpers.cc]: I weep even more for my profession.

File: src/topology/topology_location.cc [hunk 3]
```diff
@@ -0,0 +1,865 @@
...
+            // close contains objects
+            for (int i = counter; i > 0; i--) {
+                indent--;
+                for (int j = 0; j < indent; j++) {
+                    json.append ("\t");
+                }
+                json.append ("}]}\n");
+            }
+
+            json.append ("}");
```
Comment (boricj) [inline src/topology/topology_location.cc]: 😢

Comment (jeffdegott) [reply src/topology/topology_location.cc]: i agree

**Pull request merged**